### PR TITLE
Improving SequenceClassifier

### PR DESCRIPTION
**Pull request by tomkocmi**
Trochu jsem predelal sequence classifier a nasel par drobnosti (projekce linear bude brzy deprecated, tak by se mela odstranit, proto jsem ji jiz odebral z MLP a nahradil jinou metodou). 
Hlavne jsem jej otestoval, jestli se chova spravne, tak jak ma (spolecne s evaluators.accuracy).

### CODE REVIEW

File: neuralmonkey/decoders/sequence_classifier.py [hunk 1]
```diff
@@ -18,24 +19,56 @@ class SequenceClassifier(ModelPart):
...
-                 layers: Optional[List[int]]=None,
-                 activation: Callable[[tf.Tensor], tf.Tensor]=tf.tanh,
-                 dropout_keep_prob: float=0.5,
-                 save_checkpoint: Optional[str]=None,
-                 load_checkpoint: Optional[str]=None) -> None:
+                 layers: List[int] = None,
+                 activation: str = 'RELU',
+                 dropout_keep_prob: float = 0.5,
+                 save_checkpoint: Optional[str] = None,
+                 load_checkpoint: Optional[str] = None) -> None:
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_classifier.py]: tadyto má bejt rozhodně bez mezer..

Comment (tomkocmi) [inline neuralmonkey/decoders/sequence_classifier.py]: toto je hrozne divne, nikde v definici PEP8 to nemuzu najit a pokud to tam je, tak je to absurdni kdyz jinde se davaji mezery okolo rovnitek a podobnych znacek. Tak snad ted ten travis projde

Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_classifier.py]: travis na tomhle prošel i minule, chyby byly jinde. Mezi pojmenovaný parametr a jeho hodnotu se mezery kolem rovnáse nedávaj (ani v deklaraci, ani při volání funkce).

Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_classifier.py]: Podívej se na ty logy, tam je jasně napsaný, co je na který řádce v jakym souboru špatně

**Comment by tomkocmi** [comment]
Ok zjistil jsem proc 79 znaku ... je to tak nastaveny, beru to, budu to dodrzovat ... kazdopadne se me nesnazte presvedcit ze PEP8 je dobra vec ... byla dobra mozna tak 10 let zpatky ale v dnesni dobe si myslim, ze se styl programovani zasadne zmenil a tohle je uz spis sikana programatoru (hlavne ten limit na radku, kde proste v pythonu kterej hrozne indentuje pak vznika dost hnusnej kod)

**Comment by jindrahelcl** [comment]
Hlavní argument pro dodržování 80 sloupečků na řádek vidim v tom, že delší řádky bývají zbytečně komplikovaný a stejně by bylo lepší je víc rozepsat. Asi bude i něco na tom, že delší řádky se prostě hůř čtou, i když se ti na ten monitor vejdou. Mně na mým malým displeji na full screen vyjdou takhle krásně dva soubory vedle sebe a nemusej se zalamovat řádky, což je fajn.  

http://softwareengineering.stackexchange.com/questions/604/is-the-80-character-limit-still-relevant-in-times-of-widescreen-monitors

**Comment by tomkocmi** [comment]
Souhlasim, ze tobe to vyhovuje, ale kazdej chytrejsi editor ti zalamovani (a nasledne formatovani na display podle toho jestli chces dodrzovat napriklad PEP8) zvladne udelat automaticky bez fyzickych zmen v kodu, potom ale neomezujes programatory, kteri programuji jinak a vyhovuje jim vetsi limit (me se treba dva soubory vedle sebe vejdou a kazdy muze mit 120 sloupecku).

Argument ze delsi radky jsou zbytecne dlouhe neni uplne validni. Ano ve vetsine pripadu to jde, ale zrovna v pythonu, kdy indentace tridy a funkce ti sebere 8 znaku, potom dalsi dve zanoreni dalsich 8 a kvuli strukture opice tam pak das jednu funkci, ktera me vic parametru  a zacne ti z toho vylejzat neco nasledujiciho (decoder.py line 142):

```
self.decoding_b = tf.get_variable(
    "state_to_word_b", [len(self.vocabulary)],
    initializer=tf.constant_initializer(
       - math.log(len(self.vocabulary))))
```

Nebo dalsi moznost, kdy pouze vypisujes hlasku, tak si nemyslim, ze by mela byt zalamovana porad dokola a opet to muze delat editor bez modifikace zdrojaku.

**Comment by martinpopel** [comment]
Přidám se do diskuze, protože mě přesně toto tuhle zajímalo,
když jsem to řešil v https://github.com/udapi/udapi-python/blob/master/.pylintrc
a zvýšil limit na 100.
Mojí motivací je hlavně zvýšení čitelnosti kódu (jak originální).
Dokud jsem dodržoval 79, tak jsem buďto zkracoval názvy proměnných apod.,
nebo se mi kód rozlézal na moc řádek, což se mi jednak nelíbí,
jednak to zhoršuje čitelnost, protože se mi související věci nevejdou na jednu stránku.
100 mi občas taky přijde málo, ale beru ohled na lidi s menšími monitory,
kteří chtějí diffovat side by side (obojí často potřebuji sám).

V PEP8 se píše, že jednotlivé projekty mohou limit zvýšit na 100.
(Což by mohly i kdyby to tam nenapsali, že.:-)
Docstringy mají prý i tak zůstat 72, ale na to kašlu.

**Comment by tomkocmi** [comment]
Hurray :)

**Comment by varisd** [comment]
Taky se pridam do diskuze :)

Rozkouskovani dlouheho radku, na nekolik kratsich (napriklad u vyctu parametru metody) umoznuje mimo jine i lepsi sledovani zmen v ramci gitu (napri. v diffu pekne vidim, ktery parametr pribyl/ubyl/byl zmenen).

Samozrejme, limit na delku radku by sel o trochu zvysit, neznamena to, ze by kratke radky nemely stale byt preferovany. Jak psal Martin, vetsi limit se muze hodit, kdyz pouzivam delsi jmena promennych kvuli prehlednosti, nejsem pak nucen je "na silu" zkracovat.

**Comment by tomkocmi** [comment]
Argument s gitem je dobry a nejaky limit na pocet znaku by tam tedy mel byt, ale 79 znaku je opravdu malo zvlast kdyz pak vidim ten kod opice, ktera zacina byt neprehledna (viz priklad nahore) ... podle me by bylo idealni nastavit limit na 120-150 znaku s tim, ze se interne klidne snazit dodrzovat 80 napriklad parametry u funkce by mohli byt kazdy na vlastnim radku, kdyz jich je vic jak treba 5.

Co si mysli @jindrahelcl a @jlibovicky a @tomasmcz o zvyseni povoleneho limitu na radku? Za me je to nejvetsi problem, ktery zpusobuje neprehlednost kodu (ostatni moje pripominky proti PEP8 jsou spise o tom, ze otravuji programatora a zpomaluji vyvoj).

PS: pokud by tato konverzace byla na dele, navrhuji ji nekam presunout, aby se tahle branch poupravila a mergnula a nebyla tu otevrena treba tyden ci dele

**Comment by jindrahelcl** [comment]
Měli jsme tuhle diskusi cca před měsícem, ale asi sis toho nevšimnul. Před tim jsme tu diskusi měli před půl rokem. Máme tuhle diskusi pořád dokola..

Pokud máš dobrej důvod udělat dlouhou řádku, můžeš tam dát vždycky `#pylint: disable=line-too-long`
a zdůvodnit to.

Ten příklad, co dáváš, neni nepřehlednej kvůli zalamovaní řádků ale jednak kvuli tomu, že je to zalamovaný debilně, jednak kvůli tomu, že je to příliš složitej command (co do čitelnosti kódu), a taky vůbec že je to celý ohyzdný. Kdyby tam bylo tohle:

```python
bias_init_value = - math.log(len(self.vocabulary))
self.decoding_b = tf.get_variable(
    name="state_to_word_b",
    shape=[len(self.vocabulary)],
    initializer=tf.constant_initializer(bias_init_value))
```

tak je to krásně čitelný a do 80 znaků se to bohatě vejde.

A jak psal @varisd, změníš-li tady třeba název nebo bias_init_value, tak ti git diff ukáže, žes to změnil. Kdyby to bylo na jedný řádce, jen ti to řekne, že se změnila ta řádka.

Taky když máš error na řádce X, tak se ti málokdy stane, že bys nevěděl, kde na té řádce to je.

File: neuralmonkey/decoders/sequence_classifier.py [hunk 2]
```diff
@@ -24,18 +24,49 @@ def __init__(self,
                  encoders: List[Any],
                  vocabulary: Vocabulary,
                  data_id: str,
-                 layers: Optional[List[int]]=None,
-                 activation: Callable[[tf.Tensor], tf.Tensor]=tf.tanh,
+                 layers: List[int]=None,
+                 activation: str='RELU',
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_classifier.py]: proč jsi tady nenechal funkci? Teď tam musí bejt ten divnej switch..

File: neuralmonkey/decoders/sequence_classifier.py [hunk 3]
```diff
@@ -24,18 +24,49 @@ def __init__(self,
...
+            self.activation_fn = tf.nn.relu
+        elif activation.lower() == 'elu':
+            self.activation_fn = tf.nn.elu
+        elif activation.lower() == 'tanh':
+            self.activation_fn = tf.tanh
+        elif activation.lower() == 'sigmoid':
+            self.activation_fn = tf.sigmoid
+        else:
+            raise ValueError(
+                "Unknown activation function: {}".format(activation))
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_classifier.py]: Tady by prostě měla bejt rovnou ta funkce

File: neuralmonkey/decoders/sequence_classifier.py (hunk 2 above)
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_classifier.py]: jestli je to kvůli konfiguraci, tak by se to podle mě mělo řešit helper funkcí v config.utils,
která by buď dělala ten switch, nebo něco chytřejšího.

File: neuralmonkey/nn/mlp.py [hunk 4]
```diff
@@ -1,55 +1,37 @@
 # tests: lint, mypy
 
 import tensorflow as tf
-from neuralmonkey.nn.projection import linear, nonlinear
 
 
-def multilayer_perceptron(input_, layer_sizes, activation=tf.tanh,
```
Comment (jindrahelcl) [inline neuralmonkey/nn/mlp.py]: nevim, jestli se radči neměla vyhodit ta třída a tohle nechat. Tohle je novější a jednodušší kód.

File: neuralmonkey/nn/mlp.py [hunk 5]
```diff
@@ -1,55 +1,37 @@
...
         with tf.variable_scope(name):
             last_layer = mlp_input
             last_layer_size = mlp_input.get_shape()[1].value
 
             self.n_params = 0
             for i, size in enumerate(layer_configuration):
-                last_layer = tf.tanh(
-                    linear(last_layer, size,
-                           scope="dense_layer_{}".format(i + 1)))
+                last_layer = tf.contrib.layers.fully_connected(
```
Comment (jindrahelcl) [inline neuralmonkey/nn/mlp.py]: možná bych preferoval mít tyhle projekce naprogramovaný u nás - je pak přesně vidět, co se děje. ale možná je to jen moje demence. @jlibovicky?

File: neuralmonkey/decoders/sequence_classifier.py (hunk 2 above)
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_classifier.py]: beru zpět koment o konfiguraci - do konfigurace můžeš klidně rovnou napsat `activation=tf.tanh`.. Jestli nebude fungovat tf, tak by mohlo fungovat `tensorflow.tanh`.

Comment (jlibovicky) [inline neuralmonkey/decoders/sequence_classifier.py]: Jestli to nebude fungovat, tak bysme měli zařídit aby tf taky fungovalo.

Comment (tomkocmi) [inline neuralmonkey/decoders/sequence_classifier.py]: Protoze to neni konzistentni se zbytkem opice (viz prepinadlo GRU a LSTM) a lidi se museji dlabat v tensorflow, tohle prepinadlo mi prijde lepsi. Ale jestli tam chcete mit callabale tak se hadat nebudu je to celkem prkotina, me to takhle prijde cistsi

File: neuralmonkey/decoders/sequence_classifier.py (hunk 3 above)
Comment (tomkocmi) [inline neuralmonkey/decoders/sequence_classifier.py]: no a nebo udelat hybridni reseni (prepinac nebo callable) ale to se mi asi libi jeste mene nez nechani pouze callable :D

File: neuralmonkey/nn/mlp.py (hunk 5 above)
Comment (tomkocmi) [inline neuralmonkey/nn/mlp.py]: me prijde ze kusy kodu, ktery kontrolujou mraky lidi v tensowflow bychom od nich meli sourcovat ... to je hlavni duvod proc jsem to nahradil ... v puvodni verzy byla napriklad zadratovana tanh, coz nemluve o tom, ze relu je lepsi default, nepovoluje menit si tu funkci (ano dalo by se to predelat v nasi verzi, ale pak pribudou dalsi veci, ktere treba tahle fully_connected ma a nemusime vynalezat kolo)

**Comment by tomkocmi** [comment]
ok tak v tom pripade budu pouzivat #pylint: disable=line-too-long celkem casto :D ... jinak vim ze tady ta debata byla jen se resilo neco trchu jineho, ne konkretne pozadavek navysit trochu ten limit na radku ... ale asi uz to vzdavam se dohadovat, ze je to spatne nastaveno

File: neuralmonkey/decoders/sequence_classifier.py (hunk 2 above)
Comment (jlibovicky) [inline neuralmonkey/decoders/sequence_classifier.py]: Já bych byl pro to použít callable všude, kde to jenom rozumně jde. `GRUCell` a `LSTMCell` nejsou callables stejného typu (konstruktory mají argumenty) a potom se podle použité buňky ještě přepíná, co půjde jako query state do attentionu.

**Comment by jindrahelcl** [comment]
bavili jsme se o tom s @jlibovicky v Edinburghu a dohodli jsme se, že necháme těch 80 znaků na řádek (i.e. 79+newline). Klidně používej pylint disable, ale v review pak zas my budem psát, ať to předěláš, takže si asi práci neušetříš. Lidi z Heidelbergu, který aniž by se s někým z nás bavili, tak dovedli napsat kód (tuny kódu), kterej splňoval všechny naše stylový požadavky, takže bych si dovolil tvrdit, že nejsou nikterak omezující a prácizabraňující.

**Comment by jindrahelcl** [comment]
Plynulý přechod od obecné diskuse k diskusi o pull requestu: Rozchoď si u sebe lokálně ty testy, takhle jak commituješ metodou pokus/omyl je jednak zbytečný prasení toho repozitáře a jednak se tim zbytečně zaměstnává ten Travis, kterýho je jedna instance na celou organizaci, a kterej ještě k tomu běží hrozně pomalu. 

Nemusíš nutně ty testy běhat furt, ale když už ten travis zfailuje, tak si prostuduj všechny logy a pokus se opravit všechny chyby najednou, a než komitneš opravu, tak si ty testy určitě pusť u sebe.

Jindra řikal, že ti to nějak nefungovalo nebo co a že to prej kvůli tomu nepouštíš - to je chabá výmluva.. Sedíš ve vedlejším kanclu, popros ho ať se podívá, jak to pouštíš, on ti pomůže a do pěti minut to bude všechno vyřešený.

**Comment by tomkocmi** [comment]
Please do not delete my fleos branch, it is and it will still be alive for several future PR ;)

**Comment by jindrahelcl** [comment]
proč jsi to zavřel, když to ještě neni mergnutý?
